example/src/main/java/io/netty/example/dns/doh/DoHClient.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.example.dns.doh;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.dns.DefaultDnsQuery;
+import io.netty.handler.codec.dns.DefaultDnsQuestion;
+import io.netty.handler.codec.dns.DnsOpCode;
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsSection;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.ApplicationProtocolNames;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+
+public final class DoHClient {
+
+    private static final String QUERY_DOMAIN = "netty.io";
+
+    private DoHClient() {
+    }
+
+    public static void main(String[] args) throws Exception {
+        URL url = new URL("https://dns.google/dns-query");
+        EventLoopGroup group = new NioEventLoopGroup();
+
+        try {
+            final SslContext sslContext = SslContextBuilder.forClient()
+                    .applicationProtocolConfig(new ApplicationProtocolConfig(
+                            ApplicationProtocolConfig.Protocol.ALPN,
+                            ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                            ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                            ApplicationProtocolNames.HTTP_2,
+                            ApplicationProtocolNames.HTTP_1_1))
+                    .build();
+
+            Bootstrap b = new Bootstrap();
+            b.group(group)
+                    .channel(NioSocketChannel.class)
+                    .handler(new Initializer(sslContext, url));
+
+            Channel ch = b.connect(url.getHost(), url.getDefaultPort()).sync().channel();
```

Comment:
What happens here with dns.google? Is the plan use system DNS for this?

Reply:
`dns.google` does not support GET request for native DoH query. Only JSON API based GET query is supported. But for a POST request, `dns.google` works perfectly fine.

Reply:
Sorry, I meant that in order to use DoH for queries, you first rely on using system DNS.  Is there a plan here?  Or clients should configure around this in non example code.

Reply:
Nice point. Yes, if a client uses `dns.google`, it'll be resolved via system DNS that could be using native UDP/TCP.

One possible workaround would be using direct IP Address like:
```java
Channel channel = b.connect("8.8.8.8", 443).sync().channel();
```

However, let's put the worst-case scenario. A client wants to connect to "dns.google" and sends a request to ISP DNS for `dns.google` IP address. Even if ISP sends a malicious IP address (let's say 300.300.300.300, for example purpose), connection over TLS would not be successful because ISP defined IP address doesn't have a valid X.509 certificate for `dns.google`. Unless Client machine has malicious CA installed too, we can trust any DNS query over native UDP/TCP.

Reply:
This was part of why I've used vague terms like "real world", if the aim is to avoid possibly tainted DNS infrastructure, then a client (say running on an Android device) may have compromised CA certs on device (not in the app).  One of the reasons that certificate pinning exists for the actual service (non dns) HTTP requests.  

This is not a blocking topic, more raising ambiguities around why and how people would use this.  I don't have an answer here.  It's probably enough that clients use direct IP as you suggest.

Reply:
I can include the certificate pinning or strict certificate check but it will go out of context since it's more of TLS / HTTPS thing.
Clients are free to configure `SslContext`. For certificate pinning, we can have an optional configuration for this.

Reply:
More than happy for this to be out of scope, just wanted to raise it for discussion.

Reply:
Maybe we need more examples to demonstrate DoH better.